Add support for block and mutex profiling

vickenty opened the pull request:
### What does this PR do?

- add runtime config settings for golang's runtime block and mutex profiling options
- make flare collect block and mutex profiles with every performance profile
- add flare options to enable block and mutex profiling before starting cpu profile
- pass runtime block and mutex profile settings to continuous profiler when starting internal profiling

### Motivation

Get better insights into blocking behavior of the agent.

### Describe how to test your changes

1. Collect a flare from a running agent using `agent flare -p 30 -B -M`
2. Check that the flare archive contains `core-block.pprof` and `core-mutex.pprof`
3. At least block profile should have some data: `go tool pprof -top core-block.pprof`; mutex might end up empty on a lightly loaded agent
4. Restart the agent
5. Set runtime settings
   ```
   agent config set runtime_block_profile_rate 1
   agent config set runtime_mutex_profile_fraction 1
   ```
7. Collect a flare using `agent flare -p 30`
7. Flare archive should contain `core-block.pprof` and `core-mutex.pprof`.
8. Both files should have content, check using `go tool pprof -top <file>`
9. Collect a flare without access to auth token: `agent flare -p 30` (either use a non-root user or `chmod 0 /etc/datadog-agent/auth_token`)
10. Flare command should complete without errors by falling back to local mode.
9. Enable internal profiling for 10 minutes: `agent config set internal_profiling true`
10. Verify that cpu, heap, block and mutex profiles are visible in the profiling section of the app (find your host and click on the profile under the graphs)
12. Enable internal profiling in the configuration file:
    ```
    internal_profiling:
      enabled: true
      block_profile_rate 1
      mutex_profile_fraction: 1
      enable_goroutine_stacktraces: true
    ```
13. Restart the agent
14. After 10 minutes verify that cpu, heap, block, mutex and goroutine profiles are visible in the profiling section of the app (find your host and click on the profile under the graphs)

vickenty committed:
[config] add integer conversion function

will be used by two new runtime settings

vickenty committed:
[config] add runtime setting for mutex profiling

Add runtime setting to control mutex profiling fraction setting of the
go runtime

vickenty committed:
[config] add runtime setting for block profiling

vickenty committed:
[app] extract settings client init

Move most of the settings client initialization into a new public
method, so it can be used from other commands.

vickenty committed:
[flare] collect mutex and block profiles

vickenty committed:
[flare] add more profile options

Add options to enable goroutine blocking and mutex profiling before
collecting cpu profile (and to restore old values after).

Defaults chosen to minimize impact: sample 1% of mutex blocks, and
goroutine blocking longer than 10us. Based on:
https://github.com/DataDog/go-profiler-notes/blob/1f44316de777d93c8e66e4c0edd708469bf7ce5b/block.md

vickenty committed:
[profiling] integrate runtime profiling settings

Integrate runtime profiling settings and built-in continuous profiler
client. Pass current values to the continuous profiler, so it does not
overwrite current configuration if a profile type is enabled by default,
otherwise enables collection for the type in the continuous profiler if
the runtime setting is enabled.

vickenty committed:
[config] integrate runtime settings with config

Integrate new runtime profiling settings with the global configuration.
Pull initial values from config on startup, and push runtime-configured
values back, so they show up in config dump.

vickenty committed:
releasenotes

ogaca-dd commented on cmd/agent/app/flare.go:
```diff
@@ -79,6 +92,16 @@ var flareCmd = &cobra.Command{
+		return fmt.Errorf("failed to initialize settings client: %v", err)
+	}
+	prevSettings := make(map[string]interface{})
+	if err := setRuntimeProfilingSettings(c, &prevSettings); err != nil {
```
Did you consider passing `runtime_block_profile_rate` and `runtime_block_profile_rate` as parameters to the api call like `profiling` (The api handler can save the previous state and restore it when done) 
One of the advantage is that you can curl the url `http://127.0.0.1:%s/debug/pprof/...` without manually restore the profile rates.

There is no need to change anything, I am just curious if it would be possible.

vickenty replied:
Entire `/debug/pprof` is provided as-is by the pprof package, including the `profile` endpoint we call to do the profiling. We could override it, of course, that would be bit surprising. I don't think a curl user would expect the endpoint to have additional undocumented parameters.

ogaca-dd commented on cmd/agent/app/run.go:
```diff
@@ -253,6 +253,16 @@ func StartAgent() error {
 	}
 
 	// Setup Internal Profiling
+	if v := config.Datadog.GetInt("internal_profiling.block_profile_rate"); v > 0 {
```
I am not sure we need a configuration option from `datadog.yaml`. Do you have use case in mind?

vickenty replied:
If we enable internal profiling in the config, we may want to configure these optional profile types at the same time.

ogaca-dd commented on pkg/config/settings/runtime_setting_block_profile_rate.go:
```diff
@@ -35,6 +36,7 @@ func (r RuntimeBlockProfileRate) Set(value interface{}) error {
 	}
 
 	profiling.SetBlockProfileRate(rate)
+	config.Datadog.Set("internal_profiling.block_profile_rate", rate)
```
What is the reason to change the config? When you set something in `datadog.yaml` and later call `config.Datadog.Get("Something")` you expect to have the value from the config file.

vickenty replied:
So that the current values are reflected in the runtime config dump returned by `agent config`. Also for consistency with other runtime settings, which do the same thing.

ogaca-dd commented on cmd/agent/app/flare.go:
```diff
@@ -53,6 +62,10 @@ var flareCmd = &cobra.Command{
 			return fmt.Errorf("unable to set up global agent configuration: %v", err)
 		}
 
+		if err := util.SetAuthToken(); err != nil {
```
What is the reason for this change?

vickenty replied:
This is necessary to call agent api to change runtime settings.

ogaca-dd replied:
A user without an access to the auth token can now run `agent flare`, with this change it will not be possible. I am not sure this is what we want.

vickenty replied:
https://github.com/DataDog/datadog-agent/pull/8360/commits/e165515386ddef90398c5d1545ca9c3acc49a3e1

ogaca-dd commented on pkg/config/settings/runtime_setting_mutex_profile_fraction.go:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	profiling.SetMutexProfileFraction(rate)
+	config.Datadog.Set("internal_profiling.mutex_profile_fraction", rate)
```
What is the reason to change the config?

vickenty replied:
So that the current values are reflected in the runtime config dump returned by `agent config`

ogaca-dd commented on pkg/flare/archive.go:
```diff
@@ -113,6 +113,16 @@ func CreatePerformanceProfile(prefix, debugURL string, cpusec int, target *Profi
+		{
+			// mutex profile
+			Name: prefix + "-mutex.pprof",
+			URL:  debugURL + "/mutex",
```
Where this endpoint is defined?

vickenty replied:
It is provided by `pprof` package: https://golang.org/pkg/net/http/pprof/

ogaca-dd commented on pkg/flare/archive.go:
```diff
@@ -113,6 +113,16 @@ func CreatePerformanceProfile(prefix, debugURL string, cpusec int, target *Profi
+		{
+			// goroutine blocking profile
+			Name: prefix + "-block.pprof",
+			URL:  debugURL + "/block",
```
Where this endpoint is defined?

vickenty replied:
It is provided by `pprof` package: https://golang.org/pkg/net/http/pprof/

ogaca-dd reviewed (commented):
Nice PR. 
I have added several comments but minor ones.

vickenty committed:
Use correct bitWidth for int type

vickenty committed:
Fix GetInt tests on 32-bit arch

vickenty committed:
Remove useless pointer from the map

Map is passed by reference anyway

vickenty committed:
cleanup and readability improvements

Unroll getSetRuntimeSetting a bit and keep all map manipulations in
setRuntimeProfilingSetttings to improve readability.

vickenty commented on pkg/config/settings/runtime_setting.go:
```diff
@@ -95,3 +96,22 @@ func GetBool(v interface{}) (bool, error) {
+	case int:
+		return v, nil
+	case string:
+		i, err := strconv.ParseInt(v, 10, 32)
```
The idea was to use the smaller size to avoid overflow in `int(i)` later, but `ParseInt` accepts `bitWidth=0` to indicate `int` size, so that's even better.

vickenty committed:
appease linter

olivielpeau commented on cmd/agent/app/flare.go:
```diff
@@ -207,3 +231,46 @@ func createArchive(logFiles []string, pdata flare.ProfileData) (string, error) {
+	for name, value := range prev {
+		fmt.Fprintln(color.Output, color.BlueString("Restoring %s to %v", name, value))
+		if _, err := c.Set(name, fmt.Sprint(value)); err != nil {
+			fmt.Fprintln(color.Output, color.RedString("Failed restore previous value of %s: %v", name, err))
```
```suggestion
			fmt.Fprintln(color.Output, color.RedString("Failed to restore previous value of %s: %v", name, err))
```

olivielpeau commented on pkg/config/settings/runtime_setting_block_profile_rate.go:
```diff
@@ -0,0 +1,42 @@
+
+// Description returns the runtime setting's description
+func (r RuntimeBlockProfileRate) Description() string {
+	return "This setting controls the fraction of goroutine blocking events that are reported in the blocking profile"
```
Let's mention explicitly that this only affects _internal_  profiles (not application profiles that the trace-agent forwards to DD)

```suggestion
	return "This setting controls the fraction of goroutine blocking events that are reported in the internal blocking profile"
```

olivielpeau commented on pkg/config/settings/runtime_setting_mutex_profile_fraction.go:
```diff
@@ -0,0 +1,42 @@
+
+// Description returns the runtime setting's description
+func (r RuntimeMutexProfileFraction) Description() string {
+	return "This setting controls the fraction of mutex contention events that are reported in the mutex profile"
```
```suggestion
	return "This setting controls the fraction of mutex contention events that are reported in the internal mutex profile"
```

olivielpeau commented on pkg/util/profiling/profiling.go:
```diff
@@ -48,7 +49,20 @@ func Start(apiKey, site, env, service string, period time.Duration, cpuDuration
 		profiler.WithProfileTypes(profiler.CPUProfile, profiler.HeapProfile, profiler.MutexProfile),
```
Do we need to also enable the relevant profile types here to actually enable them in the DD profiler? https://docs.datadoghq.com/tracing/profiler/enabling/#pagetitle

olivielpeau replied:
additional question for @Gui774ume : `profiler.MutexProfile` was added by default in https://github.com/DataDog/datadog-agent/pull/8025, but since it's documented as having additional overhead, wouldn't it be better to only enable it if an additional config option is enabled? (which is the approach of the present PR for the core agent)

Not urgent for 7.29 since internal profiling is an internal feature, but for 7.30.

vickenty replied:
> Do we need to also enable the relevant profile types here to actually enable them in the DD profiler?
No, they will be enabled by passing `BlockProfileRate` and `MutexProfileFraction` options.

olivielpeau commented on pkg/util/profiling/runtime.go:
```diff
@@ -0,0 +1,41 @@
+func SetMutexProfileFraction(fraction int) {
+	mu.Lock()
+	defer mu.Unlock()
+	runtime.SetMutexProfileFraction(fraction)
```
Does this call only set a profiling option on the Go runtime, or does it also make the Go runtime start profiling?

(same question about `runtime.SetBlockProfileRate`)

vickenty replied:
Yes, setting these to non-zero values makes go runtime to start collecting profiling data.

olivielpeau replied:
ok, is there any risk to leaving them on "forever"? (does it use more and more memory over time for example?). If so, I think these runtime settings should be hidden (and worth a comment in the code explaining why they're hidden)

vickenty replied:
These features do allocate a small amount per-stack frame sampled. I would expect allocations to taper off with time, as the profiler accumulates location data; but there is risk that number of possible stack frames to be too large. This would cause not only increased memory usage, but also additional cpu cost (profiling data is stored in a simple fixed-size linked-list hashmap with a fairly simple hash function). Since this is a concern, I will make these settings hidden.

olivielpeau reviewed (commented):
As part of the testing and QA of this change, we should also validate that when the relevant `internal_profiling` options are enabled (through runtime settings, or through static agent config), the related profiles are visible in-app in DD profiler (https://app.datadoghq.com/profiling)

olivielpeau commented on cmd/agent/common/common.go:
```diff
@@ -68,3 +72,13 @@ func GetVersion(w http.ResponseWriter, r *http.Request) {
+		return nil, err
+	}
+	hc := util.GetClient(false)
+	return settingshttp.NewClient(hc, fmt.Sprintf("https://%v:%v/agent/config", ipcAddress, config.Datadog.GetInt("cmd_port")), "datadog-agent"), nil
```
nit: we generally refer to the core agent process as `agent` rather than `datadog-agent`

```suggestion
	return settingshttp.NewClient(hc, fmt.Sprintf("https://%v:%v/agent/config", ipcAddress, config.Datadog.GetInt("cmd_port")), "agent"), nil
```

vickenty committed:
Update cmd/agent/app/flare.go

Co-authored-by: Olivier Vielpeau <olivielpeau@users.noreply.github.com>

vickenty committed:
Update pkg/config/settings/runtime_setting_block_profile_rate.go

Co-authored-by: Olivier Vielpeau <olivielpeau@users.noreply.github.com>

vickenty committed:
Update pkg/config/settings/runtime_setting_mutex_profile_fraction.go

Co-authored-by: Olivier Vielpeau <olivielpeau@users.noreply.github.com>

vickenty committed:
Update cmd/agent/common/common.go

Co-authored-by: Olivier Vielpeau <olivielpeau@users.noreply.github.com>

vickenty committed:
fix tests for GetInt

With bitWidth=0, 2^31 is no longer prodicing an error on 64 bit arches.

olivielpeau commented:
Let's update the profiler (`dd-trace-go`) to the latest version as well while we're at it

vickenty committed:
Hide runtime profiling settings

Go runtime will start accumulating profile data as soon as this option
is set to a non-zero value. There is a risk that left on over a
prolonged period of time, it may negatively impact agent performance.

vickenty committed:
Bump dd-trace-go to 1.31.1

vickenty committed:
[profiling] add support for goroutine profiling

Can be enabled by passing an additional parameter to profiling.Start

vickenty committed:
[config] add config option to profile goroutines

Config option will be recognized in core and process-agent.

vickenty committed:
[profiling] remove deprecated option

Continuous rofiler uses trace-agent to talk to the API, and does not
need Datadog API key.

ogaca-dd commented on pkg/config/settings/runtine_setting_profiling_goroutines.go:
```diff
@@ -0,0 +1,40 @@
+		return err
+	}
+
+	config.Datadog.Set("internal_profiling.enable_goroutine_stacktraces", enabled)
```
Is `config.Datadog.Set` thread safe?

vickenty replied:
[Yes](https://github.com/DataDog/datadog-agent/blob/971434bcef8ff208132d2b723ed63923bc907cd6/pkg/config/viper.go#L34)

olivielpeau reviewed (approved):
This looks great, thanks ✨

xornivore reviewed (approved):
Awesome 🚀

vickenty committed:
[flare] don't require auth_token to start flare

It should be possible to run the flare command without having read
access to agent's auth_token file. Only initialize the settings client
and auth token if block or mutex profiles were requested and we need
them to change runtime settings.
